### ISSUE TITLE
Refresh token on OpenId not working on master branch?

### ISSUE THREAD

**Issue by jardg** [OrchardCMS/OrchardCore#1327]
Hi Im in the master branch using my own bundle which includes OrchardCore.OpenId module for securing the calls.
I get an access token and a refresh token sucessfully using my usernarme and client in the password flow with grant_type:"password" and  scope:"openid offline_access profile roles".
Then when I want to renew the token using the refresh token flow Im not able because it says "The refresh token is no longer valid."
This what it is logged during the refresh request:
```

2017-12-28 19:41:31.0776|1|Microsoft.AspNetCore.Hosting.Internal.WebHost|INFO|Request starting HTTP/1.1 POST http://localhost:2918/OrchardCore.OpenId/Access/Token application/x-www-form-urlencoded 1394 
2017-12-28 19:41:31.0864|0|OrchardCore.Modules.ModularTenantRouterMiddleware|INFO|Begin Routing Request 
2017-12-28 19:41:31.0864|0|OpenIddict.OpenIddictHandler|INFO|The token request was successfully extracted from the HTTP request: {
  "grant_type": "refresh_token",
  "client_id": "jardg",
  "client_secret": "[removed for security reasons]",
  "refresh_token": "[removed for security reasons]"
}. 
2017-12-28 19:41:31.1142|0|OpenIddict.OpenIddictHandler|INFO|The token request was successfully validated. 
2017-12-28 19:41:31.1142|0|OpenIddict.OpenIddictProvider|ERROR|The token request was rejected because the refresh token was revoked. 
2017-12-28 19:41:31.1246|0|OpenIddict.OpenIddictHandler|ERROR|The token request was rejected with the following error: invalid_grant ; The refresh token is no longer valid. 
2017-12-28 19:41:31.1246|0|OpenIddict.OpenIddictHandler|INFO|The token response was successfully returned: {
  "error": "invalid_grant",
  "error_description": "The refresh token is no longer valid."
}. 

```
Debugging I saw that FindByIdAsync method in OpenIdTokenStore gets called twice during the refresh request.First finds the token then its called again and it doesn't found it anymore because it seems it got deleted and invalidated after getting it for the first time.

Repro steps:

1. Get access and refresh token using password flow and offline scope
2. Inmediately(also happens when the access token is actually outdated) request a new access token using the refresh token in the refresh token flow

**Comment by PinpointTownes** [comment]
Hey,

In OpenIddict beta2, that was actually the expected behavior as refresh tokens used to be single-use only (a new refresh token was issued every time you sent a `grant_type=refresh_token` and the previous one was invalidated).

It's worth noting that this changed in RC1: https://github.com/openiddict/openiddict-core/issues/462. Once https://github.com/OrchardCMS/OrchardCore/pull/1286 is merged, you'll get a unique/multiple-use refresh token, which should match the behavior you were expecting.

**Comment by jersiovic** [comment]
@PinpointTownes I talked with @jardg before file this is issue and as far as I know what @jardg describes was not the expected behavior even on previous version.  Cause there is only one refresh flow request (not two in a race condition).  So,  it is not sensible that refresh flow fails processing because on a second call ON THE SAME REFRESH TOKEN FLOW REQUEST to FindByIdAsync method in OpenIdTokenStore the token doesn't exist just because it was removed previously. 
In our scenario rolling token strategy is ok,  because it is more secure and we don't have race conditions cause no SPA app refreshing tokens in parallel. Can you confirm the bug @jardg says is also fixed on next version if we configure rolling token strategy?

**Comment by sebastienros** [comment]
Just commenting to let this thing alive until I understand if this is something that needs to be triaged for beta2.

**Comment by jardg** [comment]
Hi @PinpointTownes . I just tested the new code in dev branch and its working so far except for the line:
` return _session.GetAsync<OpenIdToken>(int.Parse(identifier, CultureInfo.InvariantCulture)); ` in FindByIdAsync of OpenIdTokenStore.cs that throws parsing the identifier. I changed that line to _session.Query<OpenIdToken, OpenIdTokenIndex>(i => i.TokenId == identifier).FirstOrDefaultAsync(); and works.
The refresh flow returns me a new access token and I can get new access tokens with the same refresh token during days as you say.
Anyway is there a way of getting a new refresh token on the refresh flow or does it issue a new one when its about to expire in any flow? 
Also could the flow work the way @jersiovic  says too?I'd like to get the new access token on the refresh flow and a refresh token too invalidating the old one if possible.
Thanks for the attention.

**Comment by PinpointTownes** [comment]
Hey,

> I just tested the new code in dev branch and its working so far except for the line

This is due to the asymmetry between `FindByIdAsync` (that uses the "physical identifier") and `GetIdAsync` (that uses the new "logical identifier" introduced by Sébastien in the previous PR).

This is going to be fixed as part of https://github.com/OrchardCMS/OrchardCore/pull/1339 (in a nutshell, `FindByIdAsync` will use the logical as in the previous version and we'll have new `FindByPhysicalIdAsync` and `GetPhysicalIdAsync` methods).

> Anyway is there a way of getting a new refresh token on the refresh flow or does it issue a new one when its about to expire in any flow? Since I suppose the refresh flow isn't returning a new refresh token because the refresh token is still valid.

The new default behavior is to issue a single refresh token whose lifetime is dynamically managed, which is a good default option for the YesSql stores as we avoid persisting too many entries in the documents table.

That said, we can certainly add an option to enable the rolling tokens option (the behavior you used to have in the previous version) if it's important for you. If so, please open a new thread to track that. It shouldn't be too complicated to implement (you'd only have to set `options.UseRollingTokens = true` here: https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdConfiguration.cs#L98)).

**Comment by jersiovic** [comment]
@PinpointTownes we are evaluating both options, but before we decide if we finally use `options.UseRollingTokens` or not , one question. How exactly can be done this?:

> The new default behavior is to issue a single refresh token whose lifetime is dynamically managed

I mean with what @jardg tested on your branch with `options.UseRollingTokens = false` he found refresh flow didn't return a new refresh token, only a new access token. How is it supposed a new refresh token can be retrieved in that mode? When do you say "dynamically managed" what exactly does it mean and how does it have to be done?

**Comment by jersiovic** [comment]
I answer to myself copy pasting from here https://github.com/openiddict/openiddict-core/issues/462

> How does that work now?
By default, OpenIddict will now issue a single refresh token and manage its lifetime dynamically by extending its expiration date every time it's used in a refresh token request. This refresh token is not automatically revoked and can be used until it expires.

**Comment by jardg** [comment]
Ok. I will file a new issue for making it configurable. Does activating this option cause any incompatibility or something that should be taking into account?.Thanks.

**Comment by PinpointTownes** [comment]
> I mean when @jardg tested on your branch with options.UseRollingTokens = false he found refresh flow didn't return a new refresh token, only a new access token.

I just tested and I get a refresh token, as expected:

![image](https://user-images.githubusercontent.com/6998306/34693100-c068be7c-f4c2-11e7-99e1-65bbba209ede.png)

> Does activating this option cause any incompatibility or something that should be taking into account?

Expect the concurrency concerns mentioned by @jersiovic and the fact you'll end up storing more things in the database, nope.
